File: Sources/Core/TravelDirection.swift
import Foundation

/// 方向の推定に使う生データ。CoreLocation / CLHeading の値をそのまま入れる(Services 側で詰める)。
/// 無効値は CoreLocation の流儀に合わせて「負値」で表現される場合があるため、判定は本ファイルに集約する。
public struct MotionFix: Equatable {
    /// CLLocation.course(進行方向 0..360、負値は無効)
    public var courseDeg: Double?
    /// CLLocation.courseAccuracy(度、負値は無効。nil = 不明として扱う)
    public var courseAccuracyDeg: Double?
    /// CLLocation.speed(m/s、負値は無効)
    public var speedMps: Double?
    /// CLHeading(端末そのものの向き。ポケットに入れると進行方向と一致しない)
    public var compassHeadingDeg: Double?
    /// 位置更新からの経過秒(nil = 不明)
    public var ageSec: Double?
    /// CLLocation.horizontalAccuracy(m、負値は無効)。
    /// 方向の判定には使わないが、「そもそも位置がどれだけ確かか」を後から見るために運ぶ
    public var horizontalAccuracyM: Double?
    /// この fix を**一意に識別する時刻** [sec](CLLocation.timestamp)。
    ///
    /// なぜ要るか(2026-09-10): 頭方位は 50 Hz、GPS の fix は約 1 Hz。
    /// 受信のたびに `motionFix()` を読み直すので、**同じ fix が 50 回、独立した
    /// 証拠として学習に入っていた**。`ageSec` は「読んだ時点からの経過」なので
    /// 呼ぶたびに変わり、同一性の判定には使えない
    public var fixTime: TimeInterval?

    public init(courseDeg: Double? = nil, courseAccuracyDeg: Double? = nil,
                speedMps: Double? = nil, compassHeadingDeg: Double? = nil,
                ageSec: Double? = nil, horizontalAccuracyM: Double? = nil,
                fixTime: TimeInterval? = nil) {
        self.courseDeg = courseDeg
        self.courseAccuracyDeg = courseAccuracyDeg
        self.speedMps = speedMps
        self.compassHeadingDeg = compassHeadingDeg
        self.ageSec = ageSec
        self.horizontalAccuracyM = horizontalAccuracyM
        self.fixTime = fixTime
    }
}

/// 方向をどこから得たか。フィールドログに残して事後検証するために保持する。
public enum DirectionSource: String, Equatable {
    case course      // GPS の移動方向(歩行中のみ有効)
    case heldCourse  // 直前まで有効だった course の保持値(曲がった直後は古い向きを指す)
    case compass     // 端末の磁気コンパス(端末の向き = 進行方向とは限らない)
}

/// 直前まで有効だった course。呼び出し側が保持し、resolve に渡す
/// (Core を純粋に保つため、状態は Core の外に置く)。
public struct HeldCourse: Equatable {
    public let deg: Double
    /// 保持した時点からの経過秒
    public let ageSec: Double

    public init(deg: Double, ageSec: Double) {
        self.deg = deg
        self.ageSec = ageSec
    }
}

public struct TravelDirectionFix: Equatable {
    public let deg: Double
    public let source: DirectionSource

    public init(deg: Double, source: DirectionSource) {
        self.deg = deg
        self.source = source
    }
}

/// 「いま体はどちらを向いて進んでいるか」を決める純粋ロジック。
///
/// 設計上の決まり:
/// - iPhone をポケットに入れる前提のため、**端末コンパスは進行方向の代用にならない**。
///   歩行中は CLLocation.course(移動の軌跡から出る方向)を第一候補とする
/// - course は静止・低速時と GPS 不良時に無効になるため、その間は直前の有効な course を
///   `course_hold_sec` まで使い続ける。歩行中の向きは急には変わらないので、
///   数十秒前の移動方向のほうがポケットの中の端末の向きよりはるかに確からしい
///   (2026-08-16 の実測: コンパス退避は左右を反転させていた。docs/04 参照)
/// - それも尽きた場合のみコンパスへ退避する
///   (退避を許すかは allow_compass_fallback。既定は false)
/// - どれも使えない場合は nil。呼び出し側は「鳴らさない」または「中央で鳴らす」を選ぶ
public enum TravelDirection {
    public static func resolve(_ fix: MotionFix, held: HeldCourse? = nil,
                               params: AppParameters.Location) -> TravelDirectionFix? {
        if let course = validCourse(fix, params: params) {
            return TravelDirectionFix(deg: Geo.normalizeDeg(course), source: .course)
        }
        if let held, held.ageSec <= params.courseHoldSec {
            return TravelDirectionFix(deg: Geo.normalizeDeg(held.deg), source: .heldCourse)
        }
        if params.allowCompassFallback, let compass = fix.compassHeadingDeg, compass >= 0 {
            return TravelDirectionFix(deg: Geo.normalizeDeg(compass), source: .compass)
        }
        return nil
    }

    /// **学習・検疫へ渡してよい「いま有効な生の course」**(→ docs/13)。
    ///
    /// `resolve` と違い、**保持値もコンパス退避も通さない**。立ち止まっている間に
    /// 「止まる直前の course」と「回っている頭」を突き合わせると R が落ち、検疫が退避に落ちる
    /// — 首を回して世界固定を確かめる試験が、自分の前提を壊す(2026-09-08)。
    ///
    /// 規則を Core の 1 か所に置くのは、**呼び出し側の配線を単体テストで押さえるため**。
    /// Controller に `resolve(held: nil)` と書いていた頃は、うっかり保持値を渡す変更が入っても
    /// どのテストも落ちなかった
    public static func rawCourse(_ fix: MotionFix, params: AppParameters.Location) -> Double? {
        guard let t = resolve(fix, held: nil, params: params), t.source == .course else {
            return nil
        }
        return t.deg
    }

    /// 生の course と、**それを生んだ fix の時刻**。
    ///
    /// 学習と検疫は「同じ fix を何度読んでも証拠は 1 回ぶん」でなければならない
    /// (2026-09-10。50 Hz で同じ fix を 50 回数えていた)。時刻を一緒に運ぶことで、
    /// 呼び出し側が別々に取りに行って**取り違える**ことがなくなる
    /// **fix の時刻が無ければ渡さない。** 識別子の無い course を学習へ入れると、
    /// 同じ fix を何度でも数えられてしまう(重複排除が成立しない)
    public static func rawCourseFix(_ fix: MotionFix,
                                    params: AppParameters.Location)
        -> (deg: Double, fixTime: TimeInterval)? {
        guard let deg = rawCourse(fix, params: params), let fixTime = fix.fixTime else {
            return nil
        }
        return (deg, fixTime)
    }

    /// course が無効になった理由。ログに残して「なぜ左右が付かなかったか」を追えるようにする
    public static func rejectionReason(_ fix: MotionFix, params: AppParameters.Location) -> String? {
        guard let course = fix.courseDeg, course >= 0 else { return "course が無効" }
        guard let speed = fix.speedMps else { return "速度が不明" }
        if speed < params.minSpeedForCourseMPerS {
            return String(format: "速度不足(%.2f < %.2f m/s)", speed, params.minSpeedForCourseMPerS)
        }
        if let age = fix.ageSec, age > params.maxFixAgeSec {
            return String(format: "fix が古い(%.1f > %.1f s)", age, params.maxFixAgeSec)
        }
        if let acc = fix.courseAccuracyDeg, acc >= 0, acc > params.maxCourseAccuracyDeg {
            return String(format: "course 精度不足(%.0f > %.0f°)", acc, params.maxCourseAccuracyDeg)
        }
        return nil
    }

    /// course が「歩いている最中の、信頼できる移動方向」と言える場合のみ返す
    private static func validCourse(_ fix: MotionFix, params: AppParameters.Location) -> Double? {
        guard let course = fix.courseDeg, course >= 0 else { return nil }
        guard let speed = fix.speedMps, speed >= params.minSpeedForCourseMPerS else { return nil }
        if let age = fix.ageSec, age > params.maxFixAgeSec { return nil }
        // courseAccuracy は不明(nil)または負値(未提供)なら精度判定をしない
        if let acc = fix.courseAccuracyDeg, acc >= 0, acc > params.maxCourseAccuracyDeg { return nil }
        return course
    }
}

File: android/core/src/main/kotlin/dev/otosanpo/core/TravelDirection.kt
package dev.otosanpo.core

import kotlin.math.cos
import kotlin.math.sin

/**
 * 方向の推定に使う生データ。Android では `Location.bearing` / `bearingAccuracyDegrees` /
 * `speed` を Services 側で詰める(iOS の course / courseAccuracy / speed と 1 対 1)。
 * 無効値は負値で表される場合があるため、判定はこのファイルに集約する。
 */
data class MotionFix(
    /** 進行方向 [deg, 0..360)。負値・null は無効 */
    val courseDeg: Double? = null,
    /** 進行方向の誤差 [deg]。負値・null は不明として扱う */
    val courseAccuracyDeg: Double? = null,
    val speedMps: Double? = null,
    /** 端末そのものの向き。ポケットに入れると進行方向と一致しない */
    val compassHeadingDeg: Double? = null,
    /** 位置更新からの経過秒 */
    val ageSec: Double? = null,
    val horizontalAccuracyM: Double? = null,
)

/** 方向をどこから得たか。ログに残して事後検証する */
enum class DirectionSource(val label: String) {
    COURSE("移動方向"),
    HELD_COURSE("移動方向(保持)"),
    COMPASS("端末コンパス"),
}

/** 直前まで有効だった course。**状態は Core の外に置く**(Core を純粋に保つため) */
data class HeldCourse(val deg: Double, val ageSec: Double)

data class TravelDirectionFix(val deg: Double, val source: DirectionSource)

/**
 * 「いま体はどちらを向いて進んでいるか」を決める純粋ロジック。
 *
 * - 端末をポケットに入れる前提なので、**端末コンパスは進行方向の代用にならない**。
 *   歩行中は移動の軌跡から出る方向を第一候補にする
 * - 静止・低速・GPS 不良で無効になる間は、直前の有効な値を `courseHoldSec` まで使う。
 *   数十秒前の移動方向のほうが、ポケットの中の端末の向きよりはるかに確からしい
 * - それも尽きた場合のみコンパスへ退避(既定は false)
 * - どれも使えなければ null。呼び出し側は「鳴らさない」か「中央で鳴らす」を選ぶ
 */
object TravelDirection {
    fun resolve(fix: MotionFix, held: HeldCourse? = null,
                params: AppParameters.Location): TravelDirectionFix? {
        validCourse(fix, params)?.let {
            return TravelDirectionFix(Geo.normalizeDeg(it), DirectionSource.COURSE)
        }
        if (held != null && held.ageSec <= params.courseHoldSec) {
            return TravelDirectionFix(Geo.normalizeDeg(held.deg), DirectionSource.HELD_COURSE)
        }
        val compass = fix.compassHeadingDeg
        if (params.allowCompassFallback && compass != null && compass >= 0) {
            return TravelDirectionFix(Geo.normalizeDeg(compass), DirectionSource.COMPASS)
        }
        return null
    }

    /** course が無効になった理由。「なぜ左右が付かなかったか」を追えるようにする */
    fun rejectionReason(fix: MotionFix, params: AppParameters.Location): String? {
        val course = fix.courseDeg
        if (course == null || course < 0) return "course が無効"
        val speed = fix.speedMps ?: return "速度が不明"
        if (speed < params.minSpeedForCourseMPerS) {
            return "速度不足(%.2f < %.2f m/s)".format(speed, params.minSpeedForCourseMPerS)
        }
        val age = fix.ageSec
        if (age != null && age > params.maxFixAgeSec) {
            return "fix が古い(%.1f > %.1f s)".format(age, params.maxFixAgeSec)
        }
        val acc = fix.courseAccuracyDeg
        if (acc != null && acc >= 0 && acc > params.maxCourseAccuracyDeg) {
            return "course 精度不足(%.0f > %.0f°)".format(acc, params.maxCourseAccuracyDeg)
        }
        return null
    }

    private fun validCourse(fix: MotionFix, params: AppParameters.Location): Double? {
        val course = fix.courseDeg ?: return null
        if (course < 0) return null
        val speed = fix.speedMps ?: return null
        if (speed < params.minSpeedForCourseMPerS) return null
        val age = fix.ageSec
        if (age != null && age > params.maxFixAgeSec) return null
        val acc = fix.courseAccuracyDeg
        if (acc != null && acc >= 0 && acc > params.maxCourseAccuracyDeg) return null
        return course
    }
}

/**
 * earcon の波形を作る純粋な計算。
 *
 * **アプリと書き出しで同じ音を鳴らすために Core へ置く。**
 * Android では `AudioTrack` にそのまま流す。**iOS とまったく同じ標本列**になる。
 */
object ToneRenderer {
    /**
     * 周波数列を Hann 窓エンベロープのサイン波ブリップとして並べ、標本列にする。
     * 返すのはモノラル。定位は再生側が付ける(パン)
     */
    fun samples(tone: AppParameters.ToneSpec, sampleRate: Double, gain: Double): FloatArray {
        val blipFrames = (tone.blipSec * sampleRate).toInt()
        val gapFrames = (tone.gapSec * sampleRate).toInt()
        val count = tone.freqsHz.size
        if (count == 0 || blipFrames <= 0) return FloatArray(0)

        val out = ArrayList<Float>(count * blipFrames + maxOf(0, count - 1) * gapFrames)

        // 雑音は再現性のために自前の線形合同法で作る(同じ設定なら毎回同じ音になる)
        var seed = 0x9E3779B97F4A7C15uL
        fun nextNoise(): Double {
            seed = seed * 6_364_136_223_846_793_005uL + 1_442_695_040_888_963_407uL
            return (seed shr 11).toLong().toDouble() / (1L shl 52).toDouble() - 1.0
        }
        val mix = tone.noiseMix.coerceIn(0.0, 1.0)

        for ((i, f) in tone.freqsHz.withIndex()) {
            for (n in 0 until blipFrames) {
                val t = n / sampleRate
                val env = 0.5 * (1 - cos(2 * Math.PI * n / blipFrames))
                val tonal = sin(2 * Math.PI * f * t)
                val v = tonal * (1 - mix) + nextNoise() * mix
                out.add((v * env * gain).toFloat())
            }
            if (i < count - 1) repeat(gapFrames) { out.add(0f) }
        }
        return out.toFloatArray()
    }

    /**
     * 音色を暗くする(周波数を下げ、雑音成分を削る)。
     * HRTF では前後を判別できなかった(2026-08-18 実測)ため、背後は音色で分ける。
     * **Android はパンなので、前後を分ける手段はこれだけ。**
     */
    fun darken(tone: AppParameters.ToneSpec, darkness: Double): AppParameters.ToneSpec {
        val d = darkness.coerceIn(0.0, 1.0)
        return tone.copy(
            freqsHz = tone.freqsHz.map { it * (1 - 0.5 * d) },
            noiseMix = tone.noiseMix * (1 - d)
        )
    }
}
